collector/udpqueues_linux.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !noudp_queues
+
+package collector
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+)
+
+type (
+	udpQueuesCollector struct {
+		fs     procfs.FS
+		desc   *prometheus.Desc
+		logger log.Logger
+	}
+)
+
+func init() {
+	registerCollector("udp_queues", defaultDisabled, NewUDPqueuesCollector)
+}
+
+// NewUDPqueuesCollector returns a new Collector exposing network udp queued bytes.
+func NewUDPqueuesCollector(logger log.Logger) (Collector, error) {
+	fs, err := procfs.NewFS(*procPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open procfs: %v", err)
+	}
+	return &udpQueuesCollector{
+		fs: fs,
+		desc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "udp", "queues"),
+			"Number of allocated memory in the kernel for UDP datagrams in bytes.",
+			[]string{"queue", "ip"}, nil,
+		),
+		logger: logger,
+	}, nil
+}
+
+func (c *udpQueuesCollector) Update(ch chan<- prometheus.Metric) error {
+	s, err := c.fs.NetUDPSummary()
```

Comment:
We should allow ipv6-only systems. :grin:

Reply:
True. I added two flags which let you disable either ipv4 or ipv6. 
_Like:_ `./node_exporter --collector.udp_queues --no-collector.udp_queues.ipv4` to allow ipv6-only systems).
_Output:_
![image](https://user-images.githubusercontent.com/5091565/74953419-eca29580-5401-11ea-866e-8ee8b412603b.png)

Reply:
No, I was suggesting we auto-detect ipv4 the same way we auto-detect ipv6.

I would collect the err separately, like `errV4` and `errV6`. That way we can use the new empty data error return if neither is found.

See: https://github.com/prometheus/node_exporter/pull/1613

Reply:
Ah, got it. Should we keep the flags anyway?

Reply:
No, let's drop the flags. The less tuning knobs the better.

Reply:
Removed the knobs :-). And added new error types. The auto-detect for the specific file comes for free via prometheus/procfs lib...

Reply:
@SuperQ is this the expected error handling?